### PR TITLE
Refactor signing/assemble code in openjdk_build_pipeline.groovy to support Windows/docker builds

### PR DESCRIPTION
**Pull request by sxa**
There's a lot in here, so apologies in advance to the reviewers who will need a bit of time to review this properly ;-) Noting that there are some `SXAEC:` references in here which are things that still need to be cleaned up but the overall architecture and flow is now functional as per the builds in https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld which are now getting through all the steps and running tests.

Some things done in this PR:
- A new `batOrSh` function that runs commands under the windows batch scripts if we're on windows instead of always using `bash` as this was causing some instability (issue 3714 listed below)
- Split out some parts of buildScripts into separate functions: `buildScriptsAssemble` and `buildScriptsEclipseSigner`
- Each of those two are now listed as a separate "stage" in the pipelines, so we can see how long each part takes. The original `sign` phase has been renamed as `sign tgz/zip` to avoid ambiguity. I'm tempted to change `assemble` to `assemble & SBoM` (Noting that it also includes the `du -k` of the output from `make-adopt-build-farm.sh` which is slow - at least 10 minutes - on my current test machine)
- There are a few `openjdk_build_pipeline:` eyecatchers added so we can more easily find the separate sections while looking at the build logs (`curl -s` the `consoleText` link and grep it for `_pipeline:` - similar to the `build.sh:` lines from the temurin-build repo)
- The setting up of `envVars` has been extracted into the top level `build()` and passed into the `buildScripts()` and `buildScriptsAssemble()` functions that use it
- There is an 'ls -l' of various directories which forces shortname creation. It's a bit of a hack but the shortnames in containers is a bit hit or miss and this makes it reliable ... Without it it won't find various bits of the compiler tools
- I've left `openjdk_build_dir_arg` (obsoleted in https://github.com/adoptium/ci-jenkins-pipelines/pull/1084) commented out but I'm still tempted to just remove it now ...
- Initial stash of files for the internal signing phase now restricts the number of files copied across as `**/*` was unnecessary.

There are still some issues with the workspace cleaning options which may need to be addressed, although that can be done in subsequent PRs. e.g. `rm -rf c:/workspace/openjdk-build/cyclonedx-lib c:/workspace/openjdk-build/security` and `rm -rf ' + context.WORKSPACE + '/workspace/target/*`. Some additional work will be needed before the clean options will work. Ref: https://github.com/adoptium/infrastructure/issues/3723

Fixes https://github.com/adoptium/infrastructure/issues/3709
Supercedes https://github.com/adoptium/ci-jenkins-pipelines/pull/1103 (as this includes those changes)
Fixes https://github.com/adoptium/infrastructure/issues/3714 
Note: There is some potential follow-on work that could be done to tidy up this groovy script overall in https://github.com/adoptium/ci-jenkins-pipelines/issues/1116

## Existing windows pipeline
image

## New pipeline:
image

### CODE REVIEW

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
Ref: https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/782/
Current build with `CLEAN_WORKSPACE_BUILD_OUTPUT_ONLY_AFTER` is giving:
```
00:28:32  ===RELEASE FILE GENERATED===
00:28:32  mv: cannot stat '/cygdrive/c/workspace/openjdk-build/workspace/target//configure.txt': No such file or directory
```
https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/760/ without that optoin ran through ok

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
Possibly for later PR - reduction in the number of files copied over to the signing job to about 155 in jdk21u from over 500. I had this in initially and it has been tested as completing ok with the sign_verification job on windows.
```
// SXAEC - Windows specific 
//                                        includes: "${base_path}/hotspot/variant-server/**/*," +
//                                            "${base_path}/support/modules_cmds/**/*," +
//                                            "${base_path}/support/modules_libs/**/*," +
//                                            // JDK 16 + jpackage needs to be signed as well stash the resources folder containing the
//                                            "${base_path}/jdk/modules/jdk.jpackage/jdk/jpackage/internal/resources/*"

                                        includes: "${base_path}/hotspot/variant-server/**/*.exe," +
                                            "${base_path}/support/modules_cmds/**/*.exe," +
                                            "${base_path}/support/modules_libs/**/*.exe," +
                                            "${base_path}/hotspot/variant-server/**/*.dll," +
                                            "${base_path}/support/modules_cmds/**/*.dll," +
                                            "${base_path}/support/modules_libs/**/*.dll," +
                                            "${base_path}/hotspot/variant-server/**/*.dylib," +
                                            "${base_path}/support/modules_cmds/**/*.dylib," +
                                            "${base_path}/support/modules_libs/**/*.dylib," +
                                            "${base_path}/jdk/modules/jdk.jpackage/jdk/jpackage/internal/resources/*"

```

**Comment by sxa** [comment]
This version looks ok on windows docker and non-docker cases for versions 11 through 21 ;-)

**Comment by sxa** [comment]
run tests

**Comment by sxa** [comment]
Version | Docker | Nondocker
---|---|---
8 | [997](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/997/) | 
11 | | [#991](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/991/)
21 | [#992](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/992/) | [#990](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/990/)

**Comment by sxa** [comment]
> #  PR TESTER RESULT 
> 
> ❎ Some pipelines failed or the job was aborted! ❎ _See the `pipeline-build-check` below for more information..._

Only failure was jdk11u/linux/x64 which failed pulling the build docker image down onto a dynamically provisioned Azure machine:
```
14:23:02  Running on [build-linux-x64-2d8590](https://ci.adoptium.net/computer/build%2Dlinux%2Dx64%2D2d8590/) in /home/adoptopenjdk/workspace/build-scripts-pr-tester/build-test/jobs/jdk11u/jdk11u-linux-x64-temurin
[14:23:03  + docker pull adoptopenjdk/centos6_build_image
14:26:39  failed to register layer: write /usr/local/gcc/libexec/gcc/x86_64-pc-linux-gnu/7.5.0/cc1plus: no space left on device
```

Since jdk11u/Linux/aarch64 and Linux/x64 on other versions worked ok this should not be considered a blocker.

**Comment by sxa** [comment]
Note to self on workflow in build()
```
- context.stage('queue')
  - if (buildConfig.DOCKER_IMAGE) {
    - context.node(label)
      - if (buildConfig.DOCKER_FILE) {
        - context.docker.build(buildScripts(...))
      - else context.ws(worksapce)(context.docker.image(...) { buildScripts(...) (context.ws is windows only)
      - if (signer)
        - buildScriptsEclipseSigner()
        - call context.docker.image(DOCKER_IMAGE).assemble()
  - else // not docker
    -  if windows
      - set workspace, context.ws(workspace) { buildScripts(); buildScritpsEclipseSigner(); buildScriptsAssemble() }
    - else
      - buildScripts(); if (enableSigner) buildScriptsEclipseSigner(); buildScriptsAssemble();
if ( enableSigner ) sign(versionInfo)
if (enableInstallers) { buildInstaller(); if (enableSigner) signInstaller() }
if (enableTests) {...}
```

**Comment by sxa** [comment]
FYI @andrew-m-leonard I'm not able to verify reproducibility between the 21.0.4+7-ga and a build in docker (Note: The docker builds are using build tools instead of the full MSVC we used for the GA) It gave me 129 differences, 99.61% reproducibility. I haven't dug into what might be different in the files yet but I'll do another comparison using job 1006 from the table below once it completes as that will have been on one of the existing machines.

Build | ZIP | SBOM
---|---|---
21.0.4+7-GA | [zip](https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.4%2B7/OpenJDK21U-jdk_x64_windows_hotspot_21.0.4_7.zip) | [SBOM](https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.4%2B7/OpenJDK21U-sbom_x64_windows_hotspot_21.0.4_7.json)
windbld#999 (dockerBuild) | [zip](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/999/artifact/workspace/target/OpenJDK21U-jdk_x64_windows_hotspot_21.0.4_7.zip) | [SBOM](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/999/artifact/workspace/target/OpenJDK21U-sbom_x64_windows_hotspot_21.0.4_7.json)
windbld#1006 (not docker) | zip | SBOM

The cygwin version is different (3.4.6 vs 3.5.4). The compiler version is identical (19.37.32822) There was no `--user-openjdk-build-root-directory C:/workspace/openjdk-build/workspace/build/openjdkbuild` parameter in mine which there was on the real GA build. `BUILD_FULL_NAME` in the SBOM is `cygwin_nt-10.0-20348-x86_64--server-release` for the dockerbuild, but was `C:/workspace/openjdk-build/workspace/build/openjdkbuild` in the GA

**Comment by sxa** [comment]
Some timings for my own reference before this summary table disappears:
image

**Comment by andrew-m-leonard** [comment]
@sxa Ah of course, the July release still had the --user-openjdk-build-root-directory C:/workspace/openjdk-build/workspace/build/openjdkbuild
This will cause a difference, as the upstream fix for that was only in jdk-21.0.5.
Can you try building with the exact same path?

**Comment by sxa** [comment]
There are still some issues relating to the clean workspace options in the windows/docker case but otherwise this PR is doing what it needs to and we should be able to get it in after the release (for safety reasons)

**Comment by sxa** [comment]
> @sxa Ah of course, the July release still had the --user-openjdk-build-root-directory C:/workspace/openjdk-build/workspace/build/openjdkbuild This will cause a difference, as the upstream fix for that was only in jdk-21.0.5. Can you try building with the exact same path?

Testing at https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/1048/ (outside docker) and https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/1049/ (in a container).

In reference to our call earlier, it's worth noting that:
- https://ci.adoptium.net/job/Grinder/11190 with jdk21/nightly passed the reproducible test. The nightly build is currently 21.0.5+9
```
14:39:09  =JAVA VERSION OUTPUT BEGIN=
14:39:09  openjdk version "21.0.5-beta" 2024-10-15
14:39:09  OpenJDK Runtime Environment Temurin-21.0.5+9-202410031113 (build 21.0.5-beta+9-ea)
14:39:09  OpenJDK 64-Bit Server VM Temurin-21.0.5+9-202410031113 (build 21.0.5-beta+9-ea, mixed mode, sharing)
14:39:09  =JAVA VERSION OUTPUT END=
```
- A regrind of another run of dev.system ([G#11189](https://ci.adoptium.net/job/Grinder/11189/)) with jdk23 on a specific upstream build (version below) fails due to a mismatch in UCRT DLL on test-azure-2 (A change introduced in https://github.com/adoptium/temurin-build/pull/3981). I have initiated [G#11194](https://ci.adoptium.net/job/Grinder/11194/) on a build machine to compare
```
14:29:46  =JAVA VERSION OUTPUT BEGIN=
14:29:47  openjdk version "23.0.1-beta" 2024-10-15
14:29:47  OpenJDK Runtime Environment Temurin-23+37-202410121214 (build 23.0.1-beta+37-202410121214)
14:29:47  OpenJDK 64-Bit Server VM Temurin-23+37-202410121214 (build 23.0.1-beta+37-202410121214, mixed mode, sharing)
14:29:47  =JAVA VERSION OUTPUT END=
```
I am also running https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/windbld/1052/ which is a build of jdk21 head which is NOT marked as a release but includes the REPRODUCIBLE_COMPARE parameter so hopefully that will trigger the "old" comparison job.

**Comment by AdamBrousseau** [comment]
@sxa Thanks for the mention. I haven't had a chance to dig through this and understand it all yet. I did coincidentally run into [this issue](https://github.com/adoptium/ci-jenkins-pipelines/issues/176) again today and wondered if this would/could be addressed by this PR? If not, no worries. It doesn't seem to be a problem that often.

**Comment by sxa** [comment]
> @sxa Thanks for the mention. I haven't had a chance to dig through this and understand it all yet.

Yeah it's quite big so not trivial to review but there wasn't an easy way to break it down given all the interdependent parts and things that needed resolving. Would you be able to try out some builds in your CI with this branch (I'm testing with a copy of the build pipeline jobs) that would be good to know if it works for you.

> I did coincidentally run into [this issue](https://github.com/adoptium/ci-jenkins-pipelines/issues/176) again today and wondered if this would/could be addressed by this PR? If not, no worries.

I don't /think/ so but it may be somethign to consider when we get to working on a bit of a clean up as part of https://github.com/adoptium/ci-jenkins-pipelines/issues/1116

> It doesn't seem to be a problem that often.

I guess we could theoretically hit it on the Windows/mac platforms. We have plenty of executors on our `worker` node so not too likely to hit it in a way that causes a problem.

**Comment by sxa** [comment]
Taking this out of draft since I've had a few runs where I've been able to verify that it is building something which is binary comparable to the 21.0.5+11 GA.

**Comment by sxa** [comment]
force push to resolve perceived conflict

**Comment by sxa** [comment]
run tests

File: pipelines/build/common/openjdk_build_pipeline.groovy [hunk 1]
```diff
@@ -2024,6 +2158,9 @@ class Build {
                         context.node(label) {
                             addNodeToBuildDescription()
                             // Cannot clean workspace from inside docker container
+                            if ( buildConfig.TARGET_OS == 'windows' && buildConfig.DOCKER_IMAGE ) {
+                                context.bat('rm -rf c:/workspace/openjdk-build/cyclonedx-lib c:/workspace/openjdk-build/security')
```
Comment (andrew-m-leonard) [inline pipelines/build/common/openjdk_build_pipeline.groovy:2162]: this seems a little odd, why specific c:\workspace and why just cyclonedx-lib and security ?
presumably permissions? Why can't we just do    rm -rf ' + context.WORKSPACE + '/workspace

File: pipelines/build/common/openjdk_build_pipeline.groovy [hunk 2]
```diff
@@ -2117,38 +2256,61 @@ class Build {
...
+                                            enableSigner,
+                                            envVars
                                         )
                                     }
                                 }
+                                // Is thre potential for not enabling the signer on jdk8u instead of having this clause?
+                                if ( enableSigner && buildConfig.JAVA_TO_BUILD != 'jdk8u' ) {
+                                    context.println "openjdk_build_pipeline: running eclipse signing phase"
+                                    buildScriptsEclipseSigner()
+                                    def workspace = 'C:/workspace/openjdk-build/'
```
Comment (andrew-m-leonard) [inline pipelines/build/common/openjdk_build_pipeline.groovy:2292]: remind me... don't we build within the cloned openjdk src for Windows ?

File: pipelines/build/common/openjdk_build_pipeline.groovy (hunk 1 above)
Comment (sxa) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2162]: This was the minimal that was required to clear up the workspace to allow the `git clean` operation to succeed without failures as I recall. I would definitely like to improve the way this is done, but those two directories specifically caused a problem (potentially because they are written to during the build, and that happens inside the docker container under different permissions?) I would strongly support a separate issue being raised to look at whether we can avoid writing into that directory and/or move/change the scope of this particular `rm` operation. We do have various worksapce clean-up options on the build to clear things up, but unless that is done at the correct time (as per the comment above the bit you've highlighted) it doesn't work in the container case.

That was a longer answer that I planned, but let me know if you're ok with deferring that to an action after this PR, or if you want it addressed before you feel able to approve it. I don't like it, but I'm also conscious of the fact that the PR as-is has undergone a good amount of testing for now and there are multiple cases that would have to be tested to fully test additional changes here.

Comment (sxa) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2162]: I have done some tests without this rm and so far not reproduced the failure but I suspect it's an edge case on certain failures which triggered the value to reset the workspace so I'm still more comfortable leaving this in if possible. But definitely worth keeping an eye on so I could add some debug in to keep an eye on it and see later on if it might have failed without it (I'm running with that in another branch for the testing)

Comment (andrew-m-leonard) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2162]: ok makes sense as an edge case, but can we use:
```
'rm -rf ' + context.WORKSPACE + '/cyclonedx-lib'
'rm -rf ' + context.WORKSPACE + '/security'
```

File: pipelines/build/common/openjdk_build_pipeline.groovy (hunk 2 above)
Comment (andrew-m-leonard) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2292]: yep

File: pipelines/build/common/openjdk_build_pipeline.groovy (hunk 1 above)
Comment (sxa) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2162]: Happy to change it, but any particular reason? Just feels more efficient to spawn one shell and rm process instead of two

Comment (sxa) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2162]: At that point WORKSPACE is resolving to `c:\workspace\workspace\build-scripts\jobs\jdk21u\...` instead of the "shortname" so it would clean up the wrong directory (Also it's using `\` instead of `/` which is likely to cause problems)

Comment (sxa) [reply pipelines/build/common/openjdk_build_pipeline.groovy:2162]: OK I can get it to use that variable if I move the `workspace` definition earlier in the `build()` function and remove the multiple definitions, then wrapping this part in `context.ws(workspace)`

Comment (sxa) [reply pipelines/build/common/openjdk_build_pipeline.groovy]: Pushed the change and seems to work in the docker environment on initial tests - LMK if moving the scope of the variable is likely to cause any side-effects e.g. on other platforms..

**Comment by sxa** [comment]
run tests

File: pipelines/build/common/openjdk_build_pipeline.groovy (hunk 1 above)
Comment (andrew-m-leonard) [reply pipelines/build/common/openjdk_build_pipeline.groovy]: let me check, didn't realize this was out of scope of the context.ws()

Comment (andrew-m-leonard) [reply pipelines/build/common/openjdk_build_pipeline.groovy]: yeah looks good now, buildScripts() is run within the context.ws()...

**Review by steelhead31 — APPROVED** [review]
I've reviewed this to the best of my knowledge and ability, given the scope of change, and the extensive testing, I'm happy to approve.

**Comment by sxa** [comment]
> I've reviewed this to the best of my knowledge and ability, given the scope of change, and the extensive testing, I'm happy to approve.

Thanks - at this point I'll take that ;-) We'll find out if there are any side effects from the EA runs that'll happen overnight.

**Comment by sxa** [comment]
> ❎ Some pipelines failed or the job was aborted! ❎
> _See the `pipeline-build-check` below for more information..._

Failure was in the Alpine build where it had a connectivity problem on [dockerhost-azure-ubuntu2204-x64-1](https://ci.adoptium.net/computer/dockerhost-azure-ubuntu2204-x64-1):
```
12:43:05  Compiling 4 properties into resource bundles for jdk.httpserver
12:43:05  Compiling 11 properties into resource bundles for jdk.jartool
12:43:05  Compiling up to 71 files for COMPILE_CREATE_SYMBOLS
12:43:05  Connection attempt failed: Connection refused
12:43:05  Compiling 11 properties into resource bundles for jdk.management.agent
12:43:05  Connection attempt failed: Connection refused
12:43:05  Compiling 4 properties into resource bundles for jdk.jdi
12:43:05  Connection attempt failed: Connection refused
[...]
12:43:17  Compiling up to 3493 files for java.base
12:43:18  Connection attempt failed: Connection refused
12:43:18  Giving up
12:43:18  IOException caught during compilation: Could not connect to server after 10 attempts with timeout 4000
```
Unclear on the cause but a subsequent run on the same machine suffered a similar fate. Running a subsequent one ([build 74](https://ci.adoptium.net/job/build-scripts-pr-tester/job/build-test/job/jobs/job/jdk21u/job/jdk21u-alpine-linux-x64-temurin/74/console)) on another machine did not show a problem, and running two builds with the existing pipelines (not via the PR tester) also passed: [xlinux jdk21](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/jdk21u-linux-x64-temurin/288/) and [Alpine jdk21](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/jdk21u-alpine-linux-x64-temurin/98/). It is assumed that this is a machine specific issue and that this PR is not the cause.
